plugins/funind/recdef.ml

```diff
@@ -1416,32 +1412,26 @@ let com_terminate
     thm_name using_lemmas
     nb_args ctx
     hook =
-  let start_proof ctx (tac_start:tactic) (tac_end:tactic) =
-    let evd, env = Pfedit.get_current_context () in
-    Lemmas.start_proof thm_name
+  let start_proof env ctx (tac_start:tactic) (tac_end:tactic) =
+    let pstate = Lemmas.start_proof ~ontop:None thm_name
       (Global, false (* FIXME *), Proof Lemma) ~sign:(Environ.named_context_val env)
-      ctx (EConstr.of_constr (compute_terminate_type nb_args fonctional_ref)) ~hook;
-
-    ignore (by (Proofview.V82.tactic (observe_tac (str "starting_tac") tac_start)));
-    ignore (by (Proofview.V82.tactic (observe_tac (str "whole_start") (whole_start tac_end nb_args is_mes fonctional_ref
-    				   input_type relation rec_arg_num ))))
+      ctx (EConstr.of_constr (compute_terminate_type nb_args fonctional_ref)) ~hook in
+    let pstate = fst @@ by (Proofview.V82.tactic (observe_tac (fun _ _ -> str "starting_tac") tac_start)) pstate in
+    fst @@ by (Proofview.V82.tactic (observe_tac (fun _ _ -> str "whole_start") (whole_start tac_end nb_args is_mes fonctional_ref
+                                   input_type relation rec_arg_num ))) pstate
   in
-  start_proof ctx tclIDTAC tclIDTAC;
+  let pstate = start_proof Global.(env ()) ctx tclIDTAC tclIDTAC in
```

Comment:
I have trouble understanding if there's a semantic change here or not. It seems that previously, `start_proof` was called in a current proof context, although the original code defies my understanding. Can you comment a bit on the change made here?

Reply:
This is intentional, indeed I chose to run the declaration of the definition on a fresh proof state as doing it otherwise is broken for sure; we could keep the old behavior but it doesn't make a lot of sense IMHO.